### PR TITLE
Update @types/prettier to v2.3

### PR DESCRIPTION
**Pull request by Shinigami92**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/prettier/prettier/pull/9739/files#diff-e3117c1193bf1a00704e23c4818cbc49530cd1a3a48c6287640d0253993503e1L40>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by Shinigami92** [comment]
@ikatyang @ifiokjr @ffflorian @sosukesuzuki @kddeisz @thorn0 @fisker

Please let me now / feel free to work with me on the new types of v2.3 and provide me hints what else changed :slightly_smiling_face:

**Comment by fisker** [comment]
We added new commands to `doc`, check the history in 2021 https://github.com/prettier/prettier/commits/main/commands.md

**Comment by Shinigami92** [comment]
What are these files?

- `/types/prettier/doc.d.ts`
- `/types/prettier/OTHER_FILES.txt`

Why are these needed? Can they be removed?

**Comment by fisker** [comment]
I guess because use can `require('prettier/doc')`? IDK.

**Comment by Shinigami92** [comment]
> I guess because use can `require('prettier/doc')`? IDK.

But isn't this the same as just using `require('prettier')` and use the namespaced types and declarations?

Currently it looks like everything is maintained twice :confused:

**Comment by fisker** [comment]
Because `require('prettier')` is slower? I don't know anything about typing, but can the `Doc` import from `doc.d.ts`?

**Comment by Shinigami92** [comment]
Ah I see what you mean :sweat_smile: Havn't thought about runtime while currently working on compile time types :rofl:  
Ok I will try to move all types e.g. to `doc.d.ts` and then reuse these in `index.d.ts`

**Comment by Shinigami92** [comment]
Okay, due to these are `namespaces` it's not that easy to reexport them

So it's indeed easier to just copy the declarations over to the `doc.d.ts` :shrug:

File: types/prettier/doc.d.ts
Comment (Shinigami92) [inline types/prettier/doc.d.ts:109]: @fisker I couldn't find this in the `commands.md` :thinking: 
Is this completely removed? Or just undocumented?

Comment (fisker) [reply types/prettier/doc.d.ts:109]: Missing

File: types/prettier/doc.d.ts
Comment (fisker) [inline types/prettier/doc.d.ts:111]: This is wrong, `number | string | { type: 'root' };` is correct, I run into this few days ago.

File: types/prettier/doc.d.ts
Comment (Shinigami92) [reply types/prettier/doc.d.ts:109]: Okay so I will already add the link and you can update the `command.md` on your side? :slightly_smiling_face:

File: types/prettier/doc.d.ts
Comment (fisker) [inline types/prettier/doc.d.ts:111]: Why not the old `Align['n']`

Comment (Shinigami92) [reply types/prettier/doc.d.ts:111]: I tried to make it consistent with https://github.com/prettier/prettier/blob/main/commands.md#align, but we could change this specific case back if you want

**Comment by fisker** [comment]
FastPath renamed as AstPath

**Comment by fisker** [comment]
`AstPath#{map,each}` callback receives 3 arguments now .

File: types/prettier/index.d.ts
Comment (Shinigami92) [inline types/prettier/index.d.ts:482]: @fisker Seems the `doc` folder was renamed :thinking:  
Do we need to rename the `namespace`?

File: types/prettier/index.d.ts
Comment (Shinigami92) [inline types/prettier/index.d.ts:34]: I need to find out the types of `value`

File: types/prettier/index.d.ts
Comment (thorn0) [inline types/prettier/index.d.ts:26]: Please add an alias for backward compatibility.

Comment (Shinigami92) [reply types/prettier/index.d.ts:26]: I already thought about that, but it's a change from `2.2` to `2.3`
So if someone updates to the new types, they should just exchange their used `FastPath`
It's also a really quick search and replace

Otherwise I would just add `@deprecated` to `FastPath` and it would be marked for later removal, so why not just now directly

Comment (thorn0) [reply types/prettier/index.d.ts:26]: Adding an alias is just one line and you indeed can mark it with `@deprecated`. Please do that.

File: types/prettier/doc.d.ts
Comment (Shinigami92) [reply types/prettier/doc.d.ts:109]: https://github.com/prettier/prettier/issues/10962

File: types/prettier/index.d.ts
Comment (Shinigami92) [reply types/prettier/index.d.ts:508]: No, checked it, it's still `prettier/doc` :+1:

File: types/prettier/index.d.ts
Comment (Shinigami92) [reply types/prettier/index.d.ts:41]: Are these `value`s `T`?

Comment (fisker) [reply types/prettier/index.d.ts:37]: I think that's true in most cases, but it could be array of anything in theory.  For example:

`new AstPath({foo:[1,true,null]}).map((_,__, value) => console.log(value))`, `value` is `[1,true,null]` in this case.

File: types/prettier/parser-babel.d.ts
Comment (fisker) [inline types/prettier/parser-babel.d.ts:13]: If we are listing private parsers, a new value `__babel_estree` was added #10910, but maybe to included in v2.3 yet.

Comment (Shinigami92) [reply types/prettier/parser-babel.d.ts:13]: I have read exactly these parsers out with using `console.log(require('prettier/parser-babel'))`  
If they come out in a later version, I can add it later, okay?

File: types/prettier/index.d.ts
Comment (Shinigami92) [reply types/prettier/index.d.ts:37]: Maybe it's okay to just don't type it for now, but if someone uses these, they can improve the types on demand

**Review by fisker — APPROVED** [review]

**Review by orta — APPROVED** [review]
I'm not an expert on the prettier API (I basically just call `format` that that's it) but the original feedback from folks who are maintainers has been handled, and it looks pretty reasonable to me. 

I'm going to give it a 👍🏻  - it'd be good to give it a little time  to give others a chance to review before merging though.

**Comment by Shinigami92** [comment]
I will merge this around friday

**Comment by Shinigami92** [comment]
Ready to merge

**Pull request merged**
